pkg/cloud/openstack/openstack.go

```diff
@@ -80,3 +84,73 @@ func NewProviderAssets(config config.OperatorConfig) (common.CloudProviderAssets
 
 	return assets, nil
 }
+
+// CloudConfigTransformer implements the cloudConfigTransformer. It uses the source ConfigMap and
+// to create a new config that adheres to the configuration for CCM
+// It returns an error if the platform is not OpenStackPlatformType.
+func CloudConfigTransformer(source *corev1.ConfigMap, infra *configv1.Infrastructure) (*corev1.ConfigMap, error) {
+	if !(infra.Status.PlatformStatus != nil &&
+		infra.Status.PlatformStatus.Type == configv1.OpenStackPlatformType) {
+		return nil, fmt.Errorf("invalid platform, expected to be OpenStack")
+	}
+
+	output := source.DeepCopy()
+	infraConfigKey := infra.Spec.CloudConfig.Key
+	if val, ok := output.Data[infraConfigKey]; ok {
+		output.Data[common.DefaultConfigKey] = val
+		delete(output.Data, infraConfigKey)
+	} else {
+		return nil, fmt.Errorf(
+			"key %s specified in infra resource was not found in source configmap %s",
+			infraConfigKey, client.ObjectKeyFromObject(source),
+		)
+	}
+
+	// TODO: Do we need to care about stuff in output.BinaryData?
+	var inCfgRaw []byte
+	if v, ok := output.Data[common.DefaultConfigKey]; ok {
+		inCfgRaw = []byte(v)
+	}
+
+	cfg, err := ini.Load(inCfgRaw)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read the cloud.conf: %w", err)
+	}
+
+	// TODO: Use 'HasSection' when we bump to gopkg.in/ini.v1 v1.67 or later
+	global, err := cfg.GetSection("Global")
+	if err != nil {
+		return nil, fmt.Errorf("failed to modify the provided configuration: %w", err)
+	}
+
+	if global != nil {
+		if global.Key("secret-name").String() != "openstack-credentials" {
+			return nil, fmt.Errorf("'[Global] secret-name' is set to a non-default value")
+		}
+		if global.Key("secret-namename").String() != "kube-system" {
+			return nil, fmt.Errorf("'[Global] secret-namespace' is set to a non-default value")
+		}
+
+		// Remove the legacy keys
+		global.DeleteKey("secret-name")
+		global.DeleteKey("secret-namespace")
+		global.DeleteKey("kubeconfig-path")
+	} else {
+		// Section doesn't exist, ergo no validation to concern ourselves with
+		global, err = cfg.NewSection("Global")
+		if err != nil {
+			return nil, fmt.Errorf("failed to modify the provided configuration: %w", err)
+		}
+	}
+
+	// Add the new required ones
+	global.NewKey("use-clouds", "true")
+	global.NewKey("clouds-file", "/etc/openstack/secret/clouds.yaml")
+	global.NewKey("cloud", "openstack")
```

Comment:
Does NewKey overwrite existing keys or will it error if the key already exists? (Does it have any return values?)

Reply:
It appears to override existing keys (see [here](https://github.com/go-ini/ini/blob/v1.66.4/section.go#L65-L94)). There are return values, namely a [`Key`](https://github.com/go-ini/ini/blob/v1.66.4/key.go#L26-L39) and `error`, but we'll never hit those errors. I can add error handling here though, if needed?

Reply:
Added error handling